File: plugin/cli/src/commands/pin.ts
// plugin/cli/src/commands/pin.ts — CLI-07 sticky-override verb per ADR-011.
//
// `agentlinux pin <name>=<target>` is a STATE-ONLY mutation. It updates
// /opt/agentlinux/state/installed.d/<id>.json to record the user's intent
// about the existing install. It does NOT shell out to install.sh — the verb
// asserts something about what the existing bits mean, not a fresh install.
//
// Three target shapes:
//   - `<name>=curated`  → clears the sticky flag; the entry returns to
//                         following the catalog's curated pin on the next
//                         `upgrade`. Equivalent to "remove pin" (cf. Homebrew
//                         `brew unpin`, but recast as a positive verb).
//   - `<name>=latest`   → sticky=true + source='latest'. Upgrade --all-latest
//                         resolves the current upstream version AT UPGRADE
//                         TIME (RESEARCH Open Q4); between pin and next
//                         upgrade, sentinel.version lags behind upstream and
//                         `list` surfaces the divergence.
//   - `<name>=<semver>` → sticky=true + source='pinned' + version=<semver>.
//                         The entry is pinned at this specific version;
//                         upgrade never touches it unless --reset-all-curated.
//
// Error policy:
//   - Malformed spec (no '=', empty name, bogus target) → exit 64 (EX_USAGE).
//   - Unknown agent (not in catalog) → exit 64.
//   - Agent not installed (no sentinel) → exit 1 with clear "install first"
//     message. v0.3.0 does not support pre-declared intent on absent
//     sentinels; Phase 5+ may allow it but not this milestone.
//
// Threat model T-04-14 mitigation: parsePinSpec validates the target ∈
// {curated, latest, exact-semver} before any sentinel mutation; semver.valid
// is used (not regex) so malformed pins like '2.1' are rejected. The user
// string flows only into sentinel fields — no shell interpolation.

import semver from "semver";
import { loadCatalog } from "../catalog/loader.js";
import { detectPresence } from "../detect.js";
import { readSentinel, writeSentinel } from "../state/sentinel.js";
import type { Sentinel } from "../types.js";

export interface PinOpts {
  json?: boolean;
}

// Discriminated union — keeps call sites exhaustive via switch(parsed.target).
// 'version' carries the parsed semver string; 'curated' / 'latest' carry only
// the agent id since their version semantics are resolution-time concerns.
export type PinTarget =
  | { name: string; target: "curated" }
  | { name: string; target: "latest" }
  | { name: string; target: "version"; version: string };

export function parsePinSpec(spec: string): PinTarget {
  const eq = spec.indexOf("=");
  // eq <= 0 catches both 'no-equals' (eq === -1) and '=curated' (eq === 0,
  // empty name). Both need the usage help pointing at <name>=<target> form.
  if (eq <= 0) {
    throw new Error(
      `agentlinux pin: expected '<name>=<target>' (got '${spec}');\n  valid targets: curated, latest, or exact semver like 2.1.7`,
    );
  }
  const name = spec.slice(0, eq);
  const tgt = spec.slice(eq + 1);

  if (tgt === "curated") return { name, target: "curated" };
  if (tgt === "latest") return { name, target: "latest" };
  // semver.valid() accepts pre-releases (2.1.7-beta.1) and rejects partials
  // (2.1) + ranges (^2.1). That's the exact surface we want — pins are
  // version points, not ranges.
  if (semver.valid(tgt)) return { name, target: "version", version: tgt };

  throw new Error(
    `agentlinux pin: invalid target '${tgt}' in '${spec}';\n  valid targets: curated, latest, or exact semver (e.g. 2.1.7)`,
  );
}

export async function pinCmd(spec: string, _opts: PinOpts = {}): Promise<void> {
  let parsed: PinTarget;
  try {
    parsed = parsePinSpec(spec);
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.error(msg);
    process.exit(64); // EX_USAGE — matches install/remove error shape
    return; // defensive: process.exit is mocked to throw in tests
  }

  // Catalog lookup — validate=true because pin is a mutation path (same
  // Open Q2 rule install/upgrade follow). Reject malformed catalogs up front
  // so a pin doesn't land against a stale/broken snapshot.
  const catalog = await loadCatalog({ validate: true });
  const entry = catalog.agents.find((a) => a.id === parsed.name);
  if (!entry) {
    const available = catalog.agents
      .filter((a) => !a.test_only)
      .map((a) => a.id)
      .join(", ");
    console.error(`agentlinux: no such agent in catalog: ${parsed.name}`);
    console.error(`  available: ${available}`);
    process.exit(64);
    return;
  }

  // Sentinel must exist — pin is intent-about-existing-install, not a
  // pre-declaration. Without an install, 'pinned to 2.1.7' would be a lie
  // since no bits at 2.1.7 are on disk.
  const existing = await readSentinel(entry.id);
  if (!existing) {
    // A tool the host already has (detected, no sentinel) is `present`, not
    // absent — so route it to the SAME verb `list` recommends, not a blanket
    // "install". Three cases, mirroring the list hints:
    //   - adoptable (managed path AND in-window) → `adopt` records the bits with
    //     no reinstall.
    //   - present but NOT adoptable → `install` brings it under management:
    //     reinstall at the pin for an out-of-window managed-path tool, or relocate
    //     a non-managed-path tool (migrate). Advising `adopt` in either case would
    //     dead-end, since adopt refuses both — the exact trap QA flagged.
    const present = detectPresence(entry);
    if (present?.adoptable) {
      console.error(
        `agentlinux: ${entry.id} is present but not managed — run 'agentlinux adopt ${entry.id}' first, then pin`,
      );
    } else if (present?.canonical) {
      console.error(
        `agentlinux: ${entry.id} is present but out of the compatibility window — run 'agentlinux install ${entry.id}' to bring it under management, then pin`,
      );
    } else if (present) {
      console.error(
        `agentlinux: ${entry.id} is present at ${present.path} (not the managed path) — run 'agentlinux install ${entry.id}' to migrate it under management, then pin`,
      );
    } else {
      console.error(
        `agentlinux: ${entry.id} is not installed — run 'agentlinux install ${entry.id}' first`,
      );
    }
    process.exit(1);
    return;
  }

  // Compute next sentinel state. Spread over existing to preserve id +
  // installed_at (pin is a partial update — we only touch source/sticky/version).
  let next: Sentinel;
  switch (parsed.target) {
    case "curated":
      // Clear the override. Version stays at whatever is currently installed;
      // the next `upgrade` will classify and reconcile if divergent.
      next = { ...existing, source: "curated", sticky: false };
      console.log(`${entry.id}: pin cleared (source=curated, sticky=false)`);
      break;
    case "latest":
      // Record intent; version preserved until next `upgrade --all-latest`
      // resolves upstream (RESEARCH Open Q4).
      next = { ...existing, source: "latest", sticky: true };
      console.log(
        `${entry.id}: pinned to follow upstream latest (sticky=true); next 'upgrade --all-latest' resolves`,
      );
      break;
    case "version":
      // Exact-version pin. Record the user-asserted version on the sentinel.
      // Note: this does NOT re-install — the user is asserting the currently
      // installed bits ARE at parsed.version. If they lied, upgrade --list
      // will surface the drift on the next call.
      next = { ...existing, source: "pinned", sticky: true, version: parsed.version };
      console.log(`${entry.id}: pinned to ${parsed.version} (sticky=true)`);
      break;
  }

  await writeSentinel(next);
}

File: rust/crates/agentlinux/src/cmd/pin.rs
//! cmd/pin.rs — `agentlinux pin <spec>` (CLI-07, ADR-011).
//!
//! A STATE-ONLY sentinel
//! mutation — it NEVER dispatches a recipe. Parses the spec via the PURE
//! `parse_pin_spec` (pin_spec.rs), looks the agent up in the catalog, reads its
//! sentinel, and writes the pin target / sticky flag, printing the exact literal.
//!
//! # Guard (CLI-05)
//! The invoker guard runs ONCE in `main::dispatch` (mirroring the TS `preAction`
//! hook, index.ts:34-36) before this body — so pin does NOT re-guard. This matches
//! the TS, where `guardAgentUser` is a preAction hook, not a per-command call.
//!
//! # Exit map
//! - malformed spec → 64 (print `PinSpecError` message to stderr; the strings are
//!   ALREADY ported into `PinSpecError` — pin.ts:82 prints `err.message`).
//! - unknown agent → 64.
//! - not installed → 1 (route the present-hint sentences to STDERR via the pure
//!   `presence_gate`).

use crate::catalog::{self, FullCatalogEntry};
use crate::sentinel;
use crate::{agent_home, canonical_path, host_paths};
use agentlinux_core::detect_gates::presence_gate;
use agentlinux_core::pin_spec::{parse_pin_spec, PinTarget};
use agentlinux_core::types::CatalogEntry as CoreCatalogEntry;
use std::process::ExitCode;

const EX_USAGE: u8 = 64;

/// `agentlinux pin <spec>` body. Port of `pinCmd`.
#[must_use]
pub fn pin(spec: &str) -> ExitCode {
    // 1. Parse the spec (PURE). On error, print the message + exit 64 — the
    //  message strings are byte-identical to the TS `throw new Error(...)`.
    let parsed = match parse_pin_spec(spec) {
        Ok(p) => p,
        Err(e) => {
            crate::plog!("{e}");
            return ExitCode::from(EX_USAGE);
        }
    };

    // 2. Catalog lookup (validate:true — pin is a mutation path, pin.ts:90).
    let catalog_dir = catalog::resolve_catalog_dir();
    let agents = match catalog::load_catalog(&catalog_dir, catalog::Validate::Required) {
        Ok(a) => a,
        Err(e) => {
            crate::plog!("{e}");
            return ExitCode::from(1);
        }
    };
    let Some(entry) = catalog::find_entry(&agents, &parsed.name) else {
        return ExitCode::from(EX_USAGE);
    };

    // 3. Sentinel must exist — pin is intent-about-existing-install, not a
    //  pre-declaration.
    let existing = match sentinel::read_sentinel(&entry.id) {
        Ok(Some(s)) => s,
        Ok(None) => return pin_not_installed(entry),
        Err(e) => {
            crate::plog!("agentlinux: failed to read sentinel for {}: {e}", entry.id);
            return ExitCode::from(1);
        }
    };

    // 4. Compute the next sentinel state (partial update: preserve id +
    //  installed_at + all other fields; only source/sticky/version mutate).
    let mut next = existing;
    match parsed.target {
        PinTarget::Curated => {
            next.source = "curated".to_string();
            next.sticky = false;
            println!("{}: pin cleared (source=curated, sticky=false)", entry.id);
        }
        PinTarget::Latest => {
            next.source = "latest".to_string();
            next.sticky = true;
            println!(
                "{}: pinned to follow upstream latest (sticky=true); next 'upgrade --all-latest' resolves",
                entry.id
            );
        }
        PinTarget::Version(version) => {
            next.source = "pinned".to_string();
            next.sticky = true;
            next.version = version.clone();
            println!("{}: pinned to {version} (sticky=true)", entry.id);
        }
    }

    if let Err(e) = sentinel::write_sentinel(&next) {
        crate::plog!("agentlinux: failed to write sentinel for {}: {e}", entry.id);
        return ExitCode::from(1);
    }
    ExitCode::SUCCESS
}

/// The not-installed branch: route the present-hint sentences to
/// STDERR via the pure `presence_gate`, then exit 1.
fn pin_not_installed(entry: &FullCatalogEntry) -> ExitCode {
    let core_entry = CoreCatalogEntry::from(entry);
    let home = agent_home();
    let present = crate::cache::read_cached_agent_by_id(&entry.id).and_then(|detected| {
        presence_gate(
            &core_entry,
            &detected,
            host_paths(canonical_path(&entry.id), &home),
        )
    });
    match present {
        Some(hit) if hit.adoptable => {
            crate::plog!(
                "agentlinux: {} is present but not managed — run 'agentlinux adopt {}' first, then pin",
                entry.id, entry.id
            );
        }
        Some(hit) if hit.canonical => {
            crate::plog!(
                "agentlinux: {} is present but out of the compatibility window — run 'agentlinux install {}' to bring it under management, then pin",
                entry.id, entry.id
            );
        }
        Some(hit) => {
            crate::plog!(
                "agentlinux: {} is present at {} (not the managed path) — run 'agentlinux install {}' to migrate it under management, then pin",
                entry.id, hit.path, entry.id
            );
        }
        None => {
            crate::plog!(
                "agentlinux: {} is not installed — run 'agentlinux install {}' first",
                entry.id, entry.id
            );
        }
    }
    ExitCode::from(1)
}

#[cfg(test)]
mod pin_tests {
    use super::*;
    use crate::sentinel::Sentinel;
    use tempfile::tempdir;

    fn write_catalog(dir: &std::path::Path) {
        std::fs::write(
            dir.join("catalog.json"),
            r#"{"version":"0.3.6","agents":[
                {"id":"test-dummy","display_name":"Test","description":"d","source_kind":"script",
                 "pinned_version":"0.0.1","install_recipe_path":"install.sh",
                 "uninstall_recipe_path":"uninstall.sh","test_only":true,"tags":["test-only"]}
            ]}"#,
        )
        .unwrap();
    }

    #[test]
    fn malformed_spec_exits_64_with_ported_message() {
        // No `=` → PinSpecError::Usage. The message is byte-identical to the TS.
        assert_eq!(pin("garbage"), ExitCode::from(EX_USAGE));
        // Confirm the ported message text matches the TS throw string.
        let err = parse_pin_spec("garbage").unwrap_err();
        assert!(err.to_string().contains("expected '<name>=<target>'"));
    }

    #[test]
    fn latest_then_curated_round_trip_mutates_sentinel_only() {
        let _g = crate::test_support::env_guard();
        let cat = tempdir().unwrap();
        let state = tempdir().unwrap();
        write_catalog(cat.path());
        std::env::set_var("AGENTLINUX_CATALOG_DIR", cat.path());
        std::env::set_var("AGENTLINUX_STATE_DIR", state.path());

        // Seed an installed sentinel.
        let mut s = Sentinel::new("test-dummy".into(), "0.0.1".into(), "curated".into(), false);
        s.installed_at = Some("2026-07-28T00:00:00Z".into());
        sentinel::write_sentinel(&s).unwrap();

        // pin=latest → sticky=true, source=latest, version preserved.
        assert_eq!(pin("test-dummy=latest"), ExitCode::SUCCESS);
        let after = sentinel::read_sentinel("test-dummy").unwrap().unwrap();
        assert!(after.sticky);
        assert_eq!(after.source, "latest");
        assert_eq!(after.version, "0.0.1");
        assert_eq!(after.installed_at.as_deref(), Some("2026-07-28T00:00:00Z"));

        // pin=curated → clears sticky.
        assert_eq!(pin("test-dummy=curated"), ExitCode::SUCCESS);
        let after = sentinel::read_sentinel("test-dummy").unwrap().unwrap();
        assert!(!after.sticky);
        assert_eq!(after.source, "curated");

        // pin=<semver> → source=pinned, version updated.
        assert_eq!(pin("test-dummy=1.2.3"), ExitCode::SUCCESS);
        let after = sentinel::read_sentinel("test-dummy").unwrap().unwrap();
        assert!(after.sticky);
        assert_eq!(after.source, "pinned");
        assert_eq!(after.version, "1.2.3");

        std::env::remove_var("AGENTLINUX_CATALOG_DIR");
        std::env::remove_var("AGENTLINUX_STATE_DIR");
    }

    #[test]
    fn unknown_agent_exits_64() {
        let _g = crate::test_support::env_guard();
        let cat = tempdir().unwrap();
        write_catalog(cat.path());
        std::env::set_var("AGENTLINUX_CATALOG_DIR", cat.path());
        assert_eq!(pin("nonexistent=latest"), ExitCode::from(EX_USAGE));
        std::env::remove_var("AGENTLINUX_CATALOG_DIR");
    }

    #[test]
    fn not_installed_exits_1() {
        let _g = crate::test_support::env_guard();
        let cat = tempdir().unwrap();
        let state = tempdir().unwrap();
        write_catalog(cat.path());
        std::env::set_var("AGENTLINUX_CATALOG_DIR", cat.path());
        std::env::set_var("AGENTLINUX_STATE_DIR", state.path());
        std::env::set_var("AGENTLINUX_DETECT_CACHE", "/nonexistent/detect.json");
        // Known agent, but no sentinel and no detect cache → exit 1.
        assert_eq!(pin("test-dummy=latest"), ExitCode::from(1));
        std::env::remove_var("AGENTLINUX_CATALOG_DIR");
        std::env::remove_var("AGENTLINUX_STATE_DIR");
        std::env::remove_var("AGENTLINUX_DETECT_CACHE");
    }
}
